internal/terraform/node_resource_partial_plan.go

```diff
@@ -0,0 +1,318 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package terraform
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+)
+
+// dynamicExpandWithDeferralAllowed is a variant of dynamicExpand that we use
+// when deferred actions are enabled for the current plan.
+//
+// Once deferred actions are more stable and robust in the stacks runtime, it
+// would be nice to integrate this logic a little better with the main
+// DynamicExpand logic, but it's separate for now to minimize the risk of
+// stacks-specific behavior impacting configurations that are not opted into it.
+func (n *nodeExpandPlannableResource) dynamicExpandPartial(ctx EvalContext, knownModules []addrs.ModuleInstance, partialModules addrs.Set[addrs.PartialExpandedModule], imports addrs.Map[addrs.AbsResourceInstance, string]) (*Graph, tfdiags.Diagnostics) {
+	var g Graph
+	var diags tfdiags.Diagnostics
+
+	knownResources := addrs.MakeSet[addrs.AbsResourceInstance]()
+	partialResources := addrs.MakeSet[addrs.PartialExpandedResource]()
+	maybeOrphanResources := addrs.MakeSet[addrs.AbsResourceInstance]()
+
+	for _, moduleAddr := range knownModules {
+		resourceAddr := n.Addr.Resource.Absolute(moduleAddr)
+		resources, partials, maybeOrphans, moreDiags := n.expandKnownModule(ctx, resourceAddr, imports, &g)
+		diags = diags.Append(moreDiags)
+
+		// Track all the resources we know about.
+		knownResources = knownResources.Union(resources)
+		partialResources = partialResources.Union(partials)
+		maybeOrphanResources = maybeOrphanResources.Union(maybeOrphans)
+	}
+
+	// TODO: What about targeting and force replacement for these resources?
+	//   For now, it actually kind of works out because we don't want to make
+	//   any changes for these and that's what happens. Later, when we start
+	//   tracking deferrals in the plan, we'll need to make sure that the
+	//   targeting is applied properly.
```

Comment:
I think targeting will "technically" work, but this is going to highlight the shortcoming that `-target` already has. Targeting already causes confusion because during plan it's only guaranteed be able to target entire resources at a time, but Terraform has historically taken a best effort to prune down to the individual instance when possible. This creates some behaviors which are explained only when you understand the edges required to create the underlying graph. So not knowing which instances are available is OK, but it may create more of these confusing situations where `-target` doesn't do what the users really want.

Reply:
There's at least one case where the targeting does break which is where you have a deferred resource that depends on a non-deferred resource, and you've targeted the deferred resource. You'd expect all the parents of the deferred resource to be plan and applied as normal (since the target should include all parent changes), but during the apply step the non-deferred resource doesn't get applied because the targeted resource is missing from the plan (as it's deferred).

I've added a test that highlights this, and marked it as skipped currently because it fails. I have added a ticket to fix this later. Since we now track the deferrals in the plan, we should be able to figure out the parents of targeted resources that have been deferred so they can properly be applied during the apply stage.